### ISSUE TITLE
Eval dataset is not configured when using RL/DPO and val_set_size

### ISSUE THREAD

**Issue by filippo82** [axolotl-ai-cloud/axolotl#1226]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Setting, for example ,`val_set_size: 0.1` will configure an `eval_dataset` for the `DPOTrainer`.

### Current behaviour

The `eval_dataset` field of the returned `TrainDatasetMeta` from the `load_rl_datasets` function is None: https://github.com/filippo82/axolotl/blob/18f811978c01d567c2294140f53abcf8c086e337/src/axolotl/cli/__init__.py#L344

An `eval_dataset` is created if `test_datasets` is present in the `config.yaml`. However this parameter is not documented in the README.

### Steps to reproduce

* Start training using the config file listed below
* Add logging to [cli/train.py](https://github.com/filippo82/axolotl/blob/18f811978c01d567c2294140f53abcf8c086e337/src/axolotl/cli/train.py#L53)
* Verify that `eval_dataset=None`

### Config yaml

```yaml
base_model: teknium/OpenHermes-2.5-Mistral-7B
# base_model_config: teknium/OpenHermes-2.5-Mistral-7B
model_type: MistralForCausalLM
tokenizer_type: LlamaTokenizer
is_mistral_derived_model: true

load_in_8bit: false
# Is this set to `true` only when `adapter: qlora`?
load_in_4bit: true
strict: false

# Reinforcement Learning with DPO
rl: dpo
dpo_beta: 0.1
# max_prompt_length: 1024
# max_length: 1536

datasets:
  - path: Filippo/distilabel-intel-orca-dpo-pairs-filtered
    split: train
    type: chatml.intel
#  - path: argilla/ultrafeedback-binarized-preferences-cleaned
#    split: train_prefs
#    type: ultra_apply_chatml

dataset_prepared_path: ./last-run-prepared
output_dir: ./finetuned-model
val_set_size: 0.02
hub_model_id: Filippo/distilabeled-Hermes-2.5-Mistral-7B

sequence_len: 2048
# `sample_packing` needs to be set to `false` when fine tuning with RL
sample_packing: false
eval_sample_packing: false
pad_to_sequence_len: true
# max_packed_sequence_len: 2048

adapter: qlora
lora_model_dir:
# https://www.anyscale.com/blog/fine-tuning-llms-lora-or-full-parameter-an-in-depth-analysis-with-llama-2
lora_r: 16
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: true
lora_modules_to_save:
  - embed_tokens
  - lm_head
lora_fan_in_fan_out:
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj

num_epochs: 1
max_steps: 200
gradient_accumulation_steps: 4
micro_batch_size: 2
#optimizer: adamw_bnb_8bit
optimizer: paged_adamw_32bit
adam_beta2: 0.95
adam_epsilion: 0.00001
weight_decay: 0.0
lr_scheduler: cosine
learning_rate: 5e-5
warmup_steps: 100

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

evaluation_strategy: "steps"
# evals_per_epoch: 4
eval_steps: 10
# Default to 0
eval_table_size: 5
# Default to 128
eval_table_max_new_tokens: 128

# saves_per_epoch: 2
save_steps: 100
# Total number of checkpoints to keep
save_total_limit: 3
debug:

deepspeed:
fsdp:
fsdp_config:

#special_tokens:
#  bos_token: "<s>"
#  eos_token: "</s>"
#  unk_token: "<unk>"
special_tokens:
  bos_token: "<s>"
  eos_token: "<|im_end|>"
  unk_token: "<unk>"
  pad_token: "</s>"
tokens:
  - "<|im_start|>"
  - "<|im_end|>"

save_safetensors: true

wandb_project: distilabeled-Hermes-2.5-Mistral-7B
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:
```

### Possible solution

I guess this could be fixed by modifying the `load_split` function inside [utils/data.py](https://github.com/filippo82/axolotl/blob/18f811978c01d567c2294140f53abcf8c086e337/src/axolotl/utils/data.py#L894).

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

`axolotl` Docker image on RunPod

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by filippo82** [comment]
If you already have ideas for tackling this issue, please share them with me and I might be able to help with it and open a PR.

**Comment by PhilipMay** [comment]
Is it possible that a DPO dataset can (technicaly) not be used for evaluation and instead a SFT dataset should be used?
And that is the reason why a `val_set_size: 0.1` does and should not work for the DPO dataset?
And instead a separate `test_datasets` must be set?

**Comment by filippo82** [comment]
That's a very good question 🤔

**Comment by filippo82** [comment]
@PhilipMay I will have a look at what the Hugging Face folks do for Zephyr: https://github.com/huggingface/alignment-handbook/tree/main/recipes/zephyr-7b-beta
I am not an export so that might offer clues about this issue.

**Comment by winglian** [comment]
@filippo82 Did #1255 resolve this?

**Comment by filippo82** [comment]
Hi @winglian, thanks for the ping. Last months have been too overloaded, I'll try to look into this in the next days 🙏🏻

**Comment by PhilipMay** [comment]
Well. I actualy found a solution for this question above.
For DPO a `val_set_size` > 0 does not work at all.
Do I set `val_set_size: 0`.

If you want to have validation this dataset config works for me:

```yml
datasets:
  - ds_type: json
    split: train
    type: chatml.intel
    data_files:
      - ./data/train.jsonl.gz
test_datasets:
  - ds_type: json
    split: train
    type: chatml.intel
    data_files:
      - ./data/test.jsonl.gz      
```

**Comment by filippo82** [comment]
@winglian I still need to look into this. Looking forward to your workshop on the Maven course! I will hopefully find time to get back into fine-tuning in June. Then I should be able to check out the recent updates to `axolotl` and hopefully close this issue 🙏🏻
